src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/Ordering/AbstractAddItemCommand.cs

```diff
@@ -67,19 +68,21 @@ protected override Task<CommandStatusResult> GetCommandStatusAsync(IProjectTree
         {
             var nodeToAddTo = GetNodeToAddTo(node);
 
-            // Call OnAddingNodesAsync.
-            // Then publish changes that could have taken place in OnAddingNodesAsync.
-            await OnAddingNodesAsync(nodeToAddTo).ConfigureAwait(true);
-            await _projectTree.TreeService.PublishLatestTreeAsync(waitForFileSystemUpdates: true).ConfigureAwait(true);
-
-            var updatedNode = _projectTree.CurrentTree.Find(node.Identity);
-            var updatedNodeToAddTo = _projectTree.CurrentTree.Find(nodeToAddTo.Identity);
-            var addedNodes = OrderingHelper.GetAddedNodes(nodeToAddTo, updatedNodeToAddTo);
+            var addedElements = await OrderingHelper.AddItems(_projectVsServices.ActiveConfiguredProject, _accessor, () => OnAddingNodesAsync(nodeToAddTo)).ConfigureAwait(true);
 
-            if (addedNodes.Any())
+            if (addedElements.Any())
             {
-                // Make sure to change ConfigureAwait to "true" if we need switch back.
-                await OnAddedNodesAsync(_projectVsServices.ActiveConfiguredProject, node, addedNodes, updatedNode).ConfigureAwait(false);
+                await _accessor.OpenProjectForWriteAsync(_projectVsServices.ActiveConfiguredProject, project => OnAddedElements(project, addedElements, node, nodeToAddTo)).ConfigureAwait(true);
```

Comment:
Why ConfigureAwait(true)? Please add a comment indicating why, we usually don't want to switch back to the same thread, or back to threadpool if started on it.